src/plugins/es_ui_shared/public/request/use_request.ts

```diff
@@ -0,0 +1,162 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { useEffect, useCallback, useState, useRef, useMemo } from 'react';
+
+import { HttpSetup } from '../../../../../src/core/public';
+import {
+  sendRequest as sendStatelessRequest,
+  SendRequestConfig,
+  SendRequestResponse,
+} from './send_request';
+
+export interface UseRequestConfig extends SendRequestConfig {
+  pollIntervalMs?: number;
+  initialData?: any;
+  deserializer?: (data: any) => any;
+}
+
+export interface UseRequestResponse<D = any, E = Error> {
+  isInitialRequest: boolean;
+  isLoading: boolean;
+  error: E | null;
+  data?: D | null;
+  sendRequest: () => Promise<SendRequestResponse<D, E>>;
+}
+
+export const useRequest = <D = any, E = Error>(
+  httpClient: HttpSetup,
+  { path, method, query, body, pollIntervalMs, initialData, deserializer }: UseRequestConfig
+): UseRequestResponse<D, E> => {
+  const isMounted = useRef(false);
+
+  // Main states for tracking request status and data
+  const [error, setError] = useState<null | any>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+  const [data, setData] = useState<any>(initialData);
+
+  // Consumers can use isInitialRequest to implement a polling UX.
+  const requestCountRef = useRef<number>(0);
+  const isInitialRequest = requestCountRef.current === 0;
+  const pollIntervalIdRef = useRef<any>(null);
+
+  const clearPollInterval = useCallback(() => {
+    if (pollIntervalIdRef.current) {
+      clearTimeout(pollIntervalIdRef.current);
+      pollIntervalIdRef.current = null;
+    }
+  }, []);
+
+  // Convert our object to string to be able to compare them in our useMemo,
+  // allowing the consumer to freely passed new objects to the hook on each
+  // render without requiring them to be memoized.
+  const queryStringified = query ? JSON.stringify(query) : undefined;
+  const bodyStringified = body ? JSON.stringify(body) : undefined;
+
+  const requestBody = useMemo(() => {
+    return {
+      path,
+      method,
+      query: queryStringified ? query : undefined,
+      body: bodyStringified ? body : undefined,
+    };
+    // queryStringified and bodyStringified stand in for query and body as dependencies.
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
+  }, [path, method, queryStringified, bodyStringified]);
+
+  const sendRequest = useCallback(async () => {
+    // If we're on an interval, this allows us to reset it if the user has manually requested the
+    // data, to avoid doubled-up requests.
+    clearPollInterval();
+
+    // Pull off request ID and increment request count.
+    const requestId = requestCountRef.current++;
```

Comment:
nit: here it would be better to have 

```js
const requestId = ++requestCountRef.current;
```

and not have the `requestCountRef.current - 1;` below (the `-1` part)